op-challenger: Add actor based tests for game solver

ajsutton opened the pull request:
**Description**

Add tests for game solver where the attacker is a variety of malicious actors.  The game is played all the way through until there are no further moves to be made and then the final resolution is checked.

Also added rules to detect when the challenger posts a move that leaves it susceptible to a poisoned prestate.  One rule which tries to detect the first problematic rule currently just logs and a second rule detects when the challenger is calling step against a claim where the prestate is invalid (this call will fail logged https://github.com/ethereum-optimism/client-pod/issues/611 for the case where we try to do this) and for the case where the challenger posts a leaf claim that can be attacked using a poisoned prestate.

It also checks that none of the challenger's claims never resolved to being countered (ie the challenger always gets all its bond back).

**Tests**

So so many.

**Metadata**

- https://github.com/ethereum-optimism/client-pod/issues/103

ajsutton commented on op-challenger/game/fault/solver/rules.go:
```diff
@@ -67,13 +76,113 @@ func doNotDuplicateExistingMoves(game types.Game, action types.Action) error {
+		}
+		err = fmt.Errorf("prestate poisoned claim %v has invalid prestate and is left of honest claim countering %v at trace index %v", preStateClaim.ContractIndex, action.ParentIdx, honestTraceIndex)
+		fmt.Printf("%v\n%v\nAttack would get to %v\n", err.Error(), ancestors, pos.TraceIndex(game.MaxDepth()))
+		//return err
```
Note: Not currently returning this error to let the test continue through and see if the `detectPoisonedStep` rule catches things at the end of the game where I'm more confident in the detection logic.  I _think_ this is right though and is actually detecting the first move that leaves the challenger vulnerable.

Inphi commented on op-challenger/game/fault/solver/rules.go:
```diff
@@ -1,38 +1,47 @@
 	doNotDefendRootClaim,
+	avoidPoisonedPrestate,
+	detectPoisonedStep,
 }
```
Another good rule to add is asserting that there's at most one honest response to a claim. If, for example, the game already contains an honest attack, then the challenger should never contradict this by defending the same parent.

ajsutton replied:
Good idea, added as a game validation rule so it can utilities the bidirectional claim tree created when resolving the game to make the check simpler.

Inphi commented on op-challenger/game/fault/types/types.go:
```diff
@@ -154,14 +154,14 @@ type Claim struct {
 
 // IsRoot returns true if this claim is the root claim.
-func (c *Claim) IsRoot() bool {
+func (c Claim) IsRoot() bool {
```
nit: Making this a value receiver makes sense given that `CounteredBy` is mutable. Though none of the current receiver functions use that field. Is this more defensive programming to avoid problems later on?

ajsutton replied:
We actually use a non-pointer type for Claim nearly everywhere and it was really annoying to have to create a pointer variant just to call `IsRoot`.  But yeah it does work out nicely that because we don't use pointers often it becomes extremely hard to accidentally mutate a claim since within the challenger we never do that (only in the dispute-mon resolution code).

Inphi replied:
You can call pointer receivers using a non-pointer type. It should work without this change.
Either is fine. Though calling a value receiver copies the value for the function call. And the `Claim` struct is fat enough that I'm slightly, and _prematurely_, worried about performance calling these functions.

ajsutton replied:
hmmm, I'd swear I made this change in response to a compiler error but it seems happy to have them as pointer receivers now so I'll switch back to that.

ajsutton committed:
op-challenger: Add actor based tests for game solver

Includes rules to confirm that every move performed by the challenger would succeed as a call to contracts
Verifies that the game resolves correctly and the challenger is paid for every move it makes.
